web-ui/src/main/resources/catalog/components/viewer/projectionSwitcher/ProjectionSwitcher.js

```diff
@@ -0,0 +1,175 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
+ *
+ * Contact: Jeroen Ticheler - FAO - Viale delle Terme di Caracalla 2,
+ * Rome - Italy. email: geonetwork@osgeo.org
+ */
+
+(function() {
+  goog.provide('gn_projectionswitcher');
+
+  var module = angular.module('gn_projectionswitcher', [
+  ]);
+
+
+  /**
+   * @ngdoc directive
+   * @name gn_projectionswitcher.directive:gnProjectionSwitcher
+   * 
+   * @description The `gnProjectionSwitcher` directive adds a button to the map
+   *              that allows you to change the projection.
+   */
+  module.directive('gnProjectionSwitcher', [
+    'gnViewerSettings',
+    function(gnViewerSettings) {
+      return {
+        restrict: 'A',
+        templateUrl: '../../catalog/components/viewer/projectionSwitcher/' +
+            'partials/projectionSwitcher.html',
+        scope: {
+          map: '=gnProjectionSwitcher'
+        },
+        controllerAs: 'gnProjectionSwitcherCtrl',
+        controller: ['$scope', 'gnViewerSettings', 'gnMap', '$rootScope', 
+          function(scope, gnViewerSettings, gnMap, $rootScope) {
+
+          scope.changeLayerProjection = function(
+            layer, oldProj, newProj) {
+            if (layer instanceof ol.layer.Group) {
+              layer.getLayers().forEach(
+                  function(subLayer) {
+                    this.changeLayerProjection(
+                      subLayer, oldProj,
+                      newProj);
+                  });
+            } else if (layer instanceof ol.layer.Tile) {
+              var tileLoadFunc = layer.getSource()
+                .getTileLoadFunction();
+              layer.getSource().setTileLoadFunction(
+                tileLoadFunc);
+            } else if (layer instanceof ol.layer.Vector) {
+              var features = layer.getSource().getFeatures();
+              for (var i = 0; i < features.length; i += 1) {
+                features[i].getGeometry().transform(
+                  oldProj, newProj);
+              }
+            }
+          };
+          
+          scope.switchProjection = function(projection) {
+              var view = scope.map.getView();
+              var oldProj = view.getProjection();
+              var newProj = ol.proj.get(projection);
+              
+              if(oldProj == newProj) {
+                return;
+              }
+              
+              var projectionConfig = {};
+              
+              $.each(gnViewerSettings.mapConfig.switcherProjectionList, function(i, config) {
+                if(config['code'] == projection) {
+                  projectionConfig = config;
+                }
+              });
+              
+              var newExtent = ol.proj.transformExtent(
+                  view.calculateExtent(scope.map.getSize()), oldProj, newProj);
+              
+              if(newProj.getExtent()) {
```

Comment:
Unnecessary block; just use `newProj.getExtent()` in the config object straight away

Reply:
It may happen that the new projection does not have an extent defined on the settings. This is a failback.

Reply:
This is to help the "let's configure the map completely on the settings and avoid ows context" part.